### PR TITLE
API: Compute truncate decimal result precision based on lowest value bound

### PR DESCRIPTION
**Pull request by lliangyu-lin**
### Description
When creating a table with truncate decimal partition, the result of applying the truncate decimal may be outside the bound of decimal precision.
Ex: 
* `CREATE TABLE prod.db.sample (a DECIMAL(5,4)) USING iceberg PARTITIONED BY (TRUNCATE(10,a)`
The edge cases are -9.9991 to -9.9999. These values are legal values for the column, and the truncated result for these values are -10.0000. Since partition field type is using the same as source type (in this case `DECIMAL(5,4)`), -10.0000 does not fit into `DECIMAL(5,4)` and causing write failures and leaving orphaned data files.
* `Truncate(9, Decimal(5,4)) = Decimal(5,4)` because the lowest truncated value is the same as lowest value for `Decimal(5,4)`

The attempt fix is to compute the lowest possible truncated value (only lower bound will cause overflow), and use the max of the truncated value precision and the source type precision.

Fixes: #12915

### CODE REVIEW

**Comment by lliangyu-lin** [comment]
Hi @amogh-jahagirdar @nastra, could you help take a look on the change, if this is something we should fix?

File: api/src/main/java/org/apache/iceberg/transforms/Truncate.java [hunk 1]
```diff
@@ -513,5 +516,35 @@ public UnboundPredicate<BigDecimal> projectStrict(
       }
       return null;
     }
+
+    @Override
+    public Type getResultType(Type sourceType) {
+      if (sourceType.typeId() != Type.TypeID.DECIMAL) {
```
Comment (RussellSpitzer) [inline api/src/main/java/org/apache/iceberg/transforms/Truncate.java:522]: we would use a Precondition here for the check

File: api/src/main/java/org/apache/iceberg/transforms/Truncate.java [hunk 2]
```diff
@@ -513,5 +516,35 @@ public UnboundPredicate<BigDecimal> projectStrict(
...
+          BigDecimal.TEN
+              .pow(precision - scale)
+              .subtract(BigDecimal.ONE.movePointLeft(scale))
+              .negate();
+
+      // Truncate input bounds, find the min value that can appear in truncated output
+      BigDecimal minTrunc = minInput.divide(unit, 0, RoundingMode.FLOOR).multiply(unit);
+      int finalPrecision = Math.max(minTrunc.precision(), precision);
+
+      return Types.DecimalType.of(finalPrecision, scale);
```
Comment (RussellSpitzer) [inline api/src/main/java/org/apache/iceberg/transforms/Truncate.java:547]: A question here, if we just always just add 1 to precision will be ok? Feels like that is a bit simpler. Not sure if we have any rough edges if we do that but it would make the code consistent rather than behaving differently in certain combos.

Comment (RussellSpitzer) [reply api/src/main/java/org/apache/iceberg/transforms/Truncate.java:547]: (Either way this will require a spec change since we currently do not specify that the result type of "truncate" can be different)

File: api/src/test/java/org/apache/iceberg/transforms/TestTruncate.java [hunk 3]
```diff
@@ -85,6 +87,43 @@ public void testTruncateDecimal() {
...
+    // No widening needed
+    SerializableFunction<Object, Object> trunc99999 = Truncate.get(99999).bind(sourceType);
+    resultType = ((Transform<?, ?>) trunc99999).getResultType(sourceType);
+    assertThat(resultType).isEqualTo(sourceType);
+
+    sourceType = Types.DecimalType.of(4, 2);
+
+    // Truncate unit = 10000.00, possible truncated outputs -10000.00, 0
+    // Needs 3 extra precision widening
+    SerializableFunction<Object, Object> trunc1000000 = Truncate.get(1000000).bind(sourceType);
```
Comment (RussellSpitzer) [inline api/src/test/java/org/apache/iceberg/transforms/TestTruncate.java:122]: Should this be allowed? The truncate function is using a unit which can't be expressed in the source type?

File: api/src/main/java/org/apache/iceberg/transforms/Truncate.java (hunk 2 above)
Comment (lliangyu-lin) [reply api/src/main/java/org/apache/iceberg/transforms/Truncate.java:547]: @RussellSpitzer Thank you for the reviews! There are some extreme cases where +1 on the precision is not enough as shown in the added unit tests. Although if we restricted the input, I think +1 precision will fix the issue.

File: api/src/test/java/org/apache/iceberg/transforms/TestTruncate.java (hunk 3 above)
Comment (lliangyu-lin) [reply api/src/test/java/org/apache/iceberg/transforms/TestTruncate.java:122]: I feel this should not be allowed, but currently iceberg doesn't make any validations to input. This validation probably needs to be done in table creation level and not during the truncate computation level.
@rdblue Do you think this should be a valid case that iceberg should handle?

File: api/src/main/java/org/apache/iceberg/transforms/Truncate.java (hunk 2 above)
Comment (RussellSpitzer) [reply api/src/main/java/org/apache/iceberg/transforms/Truncate.java:547]: I wondering also as an alternative, is it worse if we just never allow the minimum truncated value to be outside of the type range? This is probably another spec change and I think precision changing is probably better (I don't think engines will actually break if we change precision unless we exceed it's inmemory type bounds ...) but I just want to explore this option.


So like if we defined
```
    Types.DecimalType sourceType = Types.DecimalType.of(5, 3);
    SerializableFunction<Object, Object> trunc1000 = Truncate.get(1000).bind(sourceType);
```
To produce values normally except when the outcome would be -100.000 and in that case we return -99.000?

The function looks a little messier there ... I'm trying to think if that's easier for engines to deal with than type changing, but my gut says no.

Comment (nandorKollar) [reply api/src/main/java/org/apache/iceberg/transforms/Truncate.java:547]: It sounds like a more general problem for me, for example over or underflows can happen for other types, like integer or long too. For example `Truncate.get(10).bind(Types.IntegerType.get()).apply(Integer.MIN_VALUE)` will underflow, instead of the expected `-2147483650` it is going to be `2147483646`. Here, underflow looks inevitable without changing the type, the result is outside of the 32 bit range.

It can also overflow, `Truncate.get(Integer.MAX_VALUE).bind(Types.IntegerType.get()).apply(1)` will result in 2 instead of 0. Note also, that the two expressions allowed by truncate behave differently in regard with overflow: `v - (v % W)` won't overflow, but `v - (((v % W) + W) % W)` can, therefore truncate can give different result in different languages. The over or underflow depends on both the value and the width.

What do you think, are these some extreme edge cases, or something possibly relevant?

File: api/src/test/java/org/apache/iceberg/transforms/TestTruncate.java (hunk 3 above)
Comment (lliangyu-lin) [reply api/src/test/java/org/apache/iceberg/transforms/TestTruncate.java:122]: Let me try to explore the option mentioned, to cap the underflow value to the lowest possible truncated value within the range. I definitely agree that having different precisions based on input isn't very consistently and elegant.

File: api/src/main/java/org/apache/iceberg/transforms/Truncate.java (hunk 2 above)
Comment (lliangyu-lin) [reply api/src/main/java/org/apache/iceberg/transforms/Truncate.java:547]: Thank you for raising the overflow / underflow issue for other types. Definitely feel like it's a general problem for truncate on numeric values. What do you think about if we just capping to lowest / highest possible values when detect there is underflow / overflow? Ex: `Truncate.get(10).bind(Types.IntegerType.get()).apply(Integer.MIN_VALUE) = Integer.MIN_VALUE`

Comment (aykut-bozkurt) [reply api/src/main/java/org/apache/iceberg/transforms/Truncate.java:547]: That feels like the solution to me. That will not mistakenly skip data files when pushing down the filters.
